### PR TITLE
`--port` is no longer mandatory for `odo url create`

### LINKED ISSUE #1772: We shouldn't require --port compulsory for `odo url create`
[kind/bug]

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the chat and talk to us if you have a question rather than a bug or feature request.

The chat room is at: https://chat.openshift.io/developers/channels/odo

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?
- Operating System: Linux
- Output of `odo version`: Master


## How did you run odo exactly?

```
/tmp/binary                                                                                                                                                                                                                                                                     ⍉
▶ odo url create myurl 
Error: Required flag(s) "port" have/has not been set
Usage:
  odo url create [component name] [flags]

Examples:
  # Create a URL for the current component with a specific port
  odo url create --port 8080

  # Create a URL with a specific name and port
  odo url create example --port 8080

  # Create a URL with a specific name by automatic detection of port (only for components which expose only one service port)
  odo url create example

  # Create a URL with a specific name and port for component frontend
  odo url create example --port 8080 --component frontend

Flags:
      --app string         Application, defaults to active application
      --component string   Component, defaults to active component.
      --context string     Use given context directory as a source for component settings
  -h, --help               Help for create
  -o, --output string      Specify output format, supported format: json
      --port int           Port number for the url of the component, required in case of components which expose more than one service port (default -1)
      --project string     Project, defaults to active project

Global Flags:
      --skip-connection-check   Skip cluster check
  -v, --v Level                 Log level for V logs. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec      Comma-separated list of pattern=N settings for file-filtered logging

 ✗  Required flag(s) "port" have/has not been set
```


## Actual behavior

Errors requiring port


## Expected behavior

Not requiring port..

It was pretty simple just to use `odo url create`. Now we have to figure out what port to actually open for the container.


## Any logs, error output, etc?

ping @girishramnani @surajnarwade @kadel

### PR DESCRIPTION
**Pull request by surajnarwade**
## What is the purpose of this change? What does it change?

* `--port` flag will no longer will be required in `odo url create`.
* `odo create` will be smart enough to get port information
* if someone mentions `odo create --port`, it will override available port

## Was the change discussed in an issue?
fixes #1772

## How to test changes?

```
odo create nodejs
odo push
odo url create 
odo push
```

### CODE REVIEW

File: pkg/odo/cli/component/create.go [hunk 1]
```diff
@@ -389,9 +389,15 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
 			return err
 		}
 		co.setResourceLimits()
+
+		var portList []string
 		if len(co.componentPorts) > 0 {
-			co.componentSettings.Ports = &(co.componentPorts)
+			portList = co.componentPorts
+		} else {
+			portList, err = co.Client.GetPortsFromBuilderImage(*co.componentSettings.Type)
```
Comment (surajnarwade) [inline pkg/odo/cli/component/create.go]: TODO: error handling

File: pkg/odo/cli/url/create.go [hunk 2]
```diff
@@ -56,7 +56,8 @@ func NewURLCreateOptions() *URLCreateOptions {
 // Complete completes UrlCreateOptions after they've been Created
 func (o *URLCreateOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
 	o.Context = genericclioptions.NewContext(cmd)
-	o.componentPort, err = url.GetValidPortNumber(o.Client, o.urlPort, o.Component(), o.Application)
+	o.localConfigInfo, err = config.NewLocalConfigInfo(o.componentContext)
```
Comment (surajnarwade) [inline pkg/odo/cli/url/create.go]: TODO: error handling

**Comment by girishramnani** [comment]
/approve
Looks good to me

File: pkg/occlient/occlient.go [hunk 3]
```diff
@@ -376,6 +376,33 @@ func imageWithMetadata(image *imagev1.Image) error {
...
+	if err != nil {
+		return []string{}, err
+	}
+	var portList []string
+	for _, po := range containerPorts {
+		port := fmt.Sprint(po.ContainerPort) + "/" + string(po.Protocol)
+		portList = append(portList, port)
+	}
+	return portList, nil
+}
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: Shouldn't we have some checks to make sure that protocol isn't blank / container port?

File: pkg/occlient/occlient.go [hunk 4]
```diff
@@ -376,6 +376,33 @@ func imageWithMetadata(image *imagev1.Image) error {
 	return nil
 }
 
+// GetPortsFromBuilderImage returns list of available port from given builder image of given component type
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: returns a list

File: pkg/occlient/occlient.go (hunk 3 above)
Comment (surajnarwade) [reply pkg/occlient/occlient.go]: * if there's no protocol, by default we will get TCP
* if ports are blank, we will get blank list

File: pkg/occlient/occlient.go (hunk 4 above)
Comment (surajnarwade) [reply pkg/occlient/occlient.go]: :+1:

**Comment by surajnarwade** [comment]
@cdrage , review needed :)

**Comment by cdrage** [comment]
Weird. Looks like tests are flaking again. 

/retest

Need us to do another review?

**Comment by cdrage** [comment]
@surajnarwade what's the status of this PR?

**Comment by surajnarwade** [comment]
@cdrage there were some flakes due to which CI was failing, which is fixed now, all :green_heart: 

need your LGTM again to merge this

File: tests/integration/json_test.go [hunk 5]
```diff
@@ -74,6 +74,16 @@ var _ = Describe("odojsonoutput", func() {
 			desiredDesCompJSON := `{"kind":"Component","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"nodejs","creationTimestamp":null},"spec":{"type":"nodejs","app": "myapp","source":"https://github.com/openshift/nodejs-ex"},"status":{"state":"Pushed"}}`
 			Expect(desiredDesCompJSON).Should(MatchJSON(actualDesCompJSON))
 
+			// odo storage create -o json
```
Comment (amitkrout) [inline tests/integration/json_test.go]: Remove this storage scenario as it is covered in the respective cmd_storage_test.go file (https://github.com/openshift/odo/blob/master/tests/integration/cmd_storage_test.go#L123)

File: tests/integration/json_test.go [hunk 6]
```diff
@@ -74,6 +74,16 @@ var _ = Describe("odojsonoutput", func() {
...
+			actualJSONStorage := helper.CmdShouldPass("odo", "storage", "create", "mystorage", "--path=/opt/app-root/src/storage/", "--size=1Gi", "-o", "json")
+			desiredJSONStorage := `{"kind":"storage","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"mystorage","creationTimestamp":null},"spec":{"size":"1Gi"},"status":{"path":"/opt/app-root/src/storage/"}}`
+			Expect(desiredJSONStorage).Should(MatchJSON(actualJSONStorage))
+
+			// odo storage list -o json
```
Comment (amitkrout) [inline tests/integration/json_test.go]: Please remove the storage list scenario.
Reason - https://github.com/openshift/odo/pull/1828/files#r302371163

**Comment by girishramnani** [comment]
looks good to me
/lgtm

**Pull request merged**